### DISCUSSION TITLE
コードレビューについて

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#25] [未分類]
コードレビューについて、つらつら書いていきます！

---

- #28 

上記のプルリクエストで、コードレビューの練習ができます！👍

上記のタスクで作成した行動規範は、GitHubから推奨されているテンプレートを流用してます。
自分で書いた部分はないので、確認作業は不要かなと思います！

なので、
良かったら皆さん、レビューの承認（LGTM）をしてみてください！😎
コードレビューの練習も兼ねて、色々触りながら確認してみてくださいーー！

**Comment by kazzyfrog** [comment]
## そもそもの、基本的な作業の流れ
1. イシューを確認して、取り組むものを決める！（自分をアサインする）
2. これから行う作業用のブランチを作成する
3. 作業してコミット&pushする
4. プルリクエストを出す
5. 誰か１人（以上）が、レビューをする
6. レビューが終われば、mainにマージする

**Reply by kazzyfrog** [reply]
このプロジェクトでは、
一旦「２人以上のレビューが必要」ということになった！

理由としては、
- プロジェクトのテーマ（作るもの）が「LGTM画像」なので、レビューに慣れないと、「LGTM」出せない。つまりLGTMステッカー使えない。。
- とはいえ、初めは１人でレビューするのは、不安？というか責任重大？に感じて、なかなかやりずらいと思うので、２人でチェックする方式にする

これで、少しずつ、レビューに慣れていければ、と思っていますー！😎
じゃないと、「LGTMステッカー」作っても、使えないので、、！⛑️

**Reply by kazzyfrog** [reply]
なお、２人が確認して、LGTMが出て、mainブランチにマージされて、
後々、バグ・表記ミスなどが発覚した場合は、仕方ないです！

そもそも、コードレビューで、バグを０にするのは無理なので、
後で修正点を見つけたら、気づいたときに修正すればOK！🙌

最低限動くかどうか、少し修正すれば改善できないかを、チェックするくらい！
以下のような感じ：

- 仕様通りに、機能が実装できているかどうか
- mainにマージする前に、最低限、表記ミスや、改善ポイントを見つける
- 自分以外の人が書いたコードを読んで、実装の意図などが分からない箇所は質問をして、学び合う場でもある👍
    - ここで、意見を交換して、より良い書き方とかを、学べることが多い！

しかも、現在は開発段階で、
実際にmainブランチにマージされたら、ユーザーが使っている製品に反映されるわけでもないので、割とカジュアルでOK！

**Comment by kazzyfrog** [comment]
こちらは、時間のある際に、
もし、深掘りして理解を深めたい場合は、読んでみてください！😎

## コードレビューについての資料：

- コードレビューについて、簡潔に記載されている：[「どうやらレビューってやつをした方が良いらしい」くらいの若手に向けた資料](https://gist.github.com/taichi/b36c135001c161aeb1b240327ed044d6)
- [実際にレビューを行う大まかな流れ](https://developer.mamezou-tech.com/blogs/2022/06/29/reviewing-pr-on-github/)

その他：
- スムーズにレビューを行うための、実践的なポイント：[まとめ記事](https://zenn.dev/s_masa/articles/956ec4c9814ddf)がある
- そして、[セルフレビュー](https://beta-chelsea.hatenadiary.jp/entry/2020/12/19/125756)も大事！

**Reply by kazzyfrog** [reply]
（少し深掘ってみる。。！）

---

## ちなみに、何を確認すればいいのか？

### 前提：
- レビュアー：レビューする人
- レビューイ：レビューされる人。つまり作業して、プルリク出した人。

### レビュアーが確認すること：
- 仕様との整合性確認：
    - 仕様通りに動作するのか、目的を達成しているのかを確認する！
    - なので、仕様を把握しておこう！
    - 作業記録があれば、さらっと見ておくと作業の内容を把握しやすい！
- プロジェクトとの整合性確認：
    - そのプロジェクト特有のコーディング規約などがあれば、それに沿っているかを確認する！
- その他、細かい表記やニュアンスの確認！

### ポイント:
レビューイ（レビューされる人）は、:
- 作業過程を記録しよう。（プルリク内に、メモスレッドの`#番号`を貼っておくと親切！）
- プルリクエストで、作業の内容を簡潔に伝えよう
- それらを実践することで、あなたの行った作業を把握・確認するのがめっちゃ簡単になる！つまりレビューがサクサク進む！

レビュアーは、:
- 率先的に、自分以外の人の作業を把握しよう
    - `main`ブランチのコードは全員のモノ！作業するだけじゃなく、チェックすることも大事！
- 言葉使いに気をつけよう！
    - レビューコメントは間違いの指摘が多くなってしまうので、そんな気持ちは無いのにコメントがきつくなってしまう場合がある！
    - テキストコミュニケーションなので、ニュアンスが伝わりづらい分、フランクに話し合うようにするのが大切！
- 作業記録がなく、プルリクエストの文章もわかりづらい場合、作業内容を把握できないことを素直を伝えよう！
    - レビューの負荷が高すぎる（丸投げ的な）場合、先延ばし・スルーされがちです。
    - 「よくわからないけど、、この人のプルリクだから、、たぶん、LGTM👍」とするよりは、正直にわからないと伝えた方が、全然良いです！
    - レビューでは、ちゃんと話し合いながら、レビューイ・レビュアー共に改善し、高め合う機会にすることが重要！

あとは、全員がレビューイ・レビュアーどちらの役割もこなしていく中で、
様々な工夫を取り入れて、改善していければいいかなと思ってますー👍

**Comment by kazzyfrog** [comment]
## コードレビューの流れ

### 1.プルリクエスト（PR）を開く！

<img width="1433" alt="スクリーンショット 2024-08-07 0 14 17" src="https://github.com/user-attachments/assets/0577ed6d-b5d1-4a0b-96eb-a246d1cd9589">

**Reply by kazzyfrog** [reply]
### 2.PRの画面から、「Files Changed」というタブから、レビュー画面に移動する！
<img width="857" alt="スクリーンショット 2024-08-07 0 15 24" src="https://github.com/user-attachments/assets/d1de7b14-daae-463b-a09d-298053532392">

**Reply by kazzyfrog** [reply]
### 3.修正依頼や、実装の意図を質問したい箇所があれば、行数を選択し、コメントを追加する！
ポイント：
- レビューコメントを追加したら、「Start a review」ボタンを押す！
- レビューコメントは、気になる箇所が複数ある場合、複数追加できます！
<img width="1440" alt="スクリーンショット 2024-08-07 0 18 00" src="https://github.com/user-attachments/assets/ed87f54a-f5f4-43c8-9b67-3f0fd67e0cc2">

**Reply by kazzyfrog** [reply]
### 4.一通りレビューが終わったら、

1. 右側の「Review changes」という、緑のボタンを押す
2. そのレビューの内容にチェックを入れ、適宜コメントを追加する
    - Comment：コメントのみ送信する
    - Approve：問題ないので、承認する
    - Request Changes：変更や質問対応などをお願いする
3. 「Submit Review」という緑のボタンで、内容を送信して、完了！

もちろん、上記は後から編集など可能です！👍

<img width="1440" alt="スクリーンショット 2024-08-07 0 16 15" src="https://github.com/user-attachments/assets/3a41f688-d02b-4f87-8ae8-6288b6d81eed">

**Reply by kazzyfrog** [reply]
## 5.LGTM、、？🫠🫠🫠
もし、修正が必要なかったり、もう修正を終えて、問題ないと判断した場合は、
「Approve」にチェックを入れて、
### LGTM
というコメントをするのが、開発者の一般的な慣習です！

**Reply by kazzyfrog** [reply]
将来的には、ここでオリジナルの画像（ステッカー）を貼る感じです！！！！！

テンション上がる！🤩

**Reply by siso001** [reply]
楽しみぃ！！！

**Reply by kagomen** [reply]
なんとなく「こうかな？」でレビューしましたが、こうやってちゃんと手順書？で確認できると安心しますね……！
まとめありがとうございます！！🥳
